File: app/src/main/java/com/example/uscatterbrain/network/wifidirect/WifiDirectProviderImpl.java
package com.example.uscatterbrain.network.wifidirect;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class WifiDirectProviderImpl implements WifiDirectProvider {

    private final WifiDirectRadioModule wifiDirectRadioModule;

    @Inject
    public WifiDirectProviderImpl(
        WifiDirectRadioModule wifiDirectRadioModule

    ) {
        this.wifiDirectRadioModule = wifiDirectRadioModule;
    }

    @Override
    public WifiDirectRadioModule getRadioModule() {
        return wifiDirectRadioModule;
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/WifiDirectProviderImpl.kt
package net.ballmerlabs.uscatterbrain

import android.content.Context
import android.net.wifi.p2p.WifiP2pManager
import java.util.concurrent.atomic.AtomicReference
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WifiDirectProviderImpl @Inject constructor(
    private val ctx: Context,
) : WifiDirectProvider {
    private val channel = AtomicReference<WifiP2pManager.Channel?>(null)
    override fun getChannel(): WifiP2pManager.Channel? {
        return channel.updateAndGet { v ->
            when (v) {
                null -> getManager()?.initialize(ctx, ctx.mainLooper, null)
                else -> v
            }
        }
    }

    override fun getManager(): WifiP2pManager? {
        return ctx.getSystemService(Context.WIFI_P2P_SERVICE) as WifiP2pManager?
    }
}